### ISSUE TITLE
Does ganesha plan to support state persistence ?

### ISSUE THREAD

**Issue by liyinshubyte** [nfs-ganesha/nfs-ganesha#816]
Currently, when ganesha restart, the lock state on ganesha is lost, so does ganesha plan to support state persistence ?

**Comment by ffilz** [comment]
That would be pretty intensive use of persistent storage to record every lock and unlock. The state would also have to be reconstituted in a way that nothing depended on pointers. Every single NFS v4.1 operation would result in at least one extra storage operation (to record the sequence number update), and really two because you would also need to persist the slot cache. And there might be additional storage transactions to make the whole thing transactional.

**Comment by liyinshubyte** [comment]
@ffilz Thanks the reply, we plan to support state persistence. Is there any discussion before ? so that we can consider more.

**Comment by ffilz** [comment]
There really hasn't been much discussion other than raising the topic, and brief conversations about what is necessary to accomplish state persistence. Are you doing this for both NFSv3 and NFSv4.x?

To persist state you will need:

For NFSv3 - you will need to re-work how rpc.statd works so that it maintains the active clients as active after reboot as opposed to putting them in recovery, and you will need to maintain the NSM generation number so all state continues to be associated with the original generation. You will need to make sure SM_NOTIFY messages DO NOT go to the clients.

For NFSv4.x you will need to preserve all the information necessary to continue using the existing short form client ids and stateids including the server's epoch. This includes the last response for each stateid for NFSv4.0 and the slot cache for NFSv4.1+.

You will need to maintain the open/share and lock databases, along with the delegation if you want to try and support those, and layouts if you will be using pNFS. Note that some of the share information is maintained by the FSAL.

All of this will require major changes to the Ganesha code from protocol layer, through SAL, and down to the FSAL. These changes will almost certainly require forking Ganesha.

You should engage the wider NFS community to see what other efforts to provide state persistence have been made and what challenges and successes have been encountered.

This is a very major undertaking, even if you limit yourself to NFSv4.1+ and I wish you the best of luck.

**Comment by liyinshubyte** [comment]
@ffilz Thanks the reply, we will first work on NFSv4.1+, and first support lock persistence.

**Comment by ffilz** [comment]
It seems like there may be more interest in persistent state and maybe my stance yesterday was a bit strong.

Persistent state IS going to require significant code changes and increase in complexity. Ideally, we would have at least an in-tree FSAL, preferably one we can run, that can support persistent state (It COULD just be a FSAL_VFS hack that stores the state in an SQL-Lite database or something purely as a proof of concept and correctness) so we can test the code and keep it maintained.

**Comment by liyinshubyte** [comment]
@ffilz Thanks the reply. If we first support lock persistence, is it much easier ?

**Comment by ffilz** [comment]
Not really, at least if you are doing NFSv4.,x since an NFSv4.x lock is dependent on an open stateid and both are dependent on a clientid (which for NFSv4.1 also means persistent sessions). Otherwise the NFSv4.x client would have no way of leveraging the persisted lock.

Now there might be an intermediate stage where a cluster maintains enough knowledge of the locks that when a failure occurs, the locks are never released, and instead, when the client goes through the recovery process, the locks are seamlessly re-acquired by the client. But the clients would still be exposed to grace period and state recovery. As long as the backend can re-associate the lockowner of the reclaim with the preserved lock AND the backend has a mechanism for releasing locks that are not reclaimed by the end of the grace period then a real benefit should result. This is effectively the solution FSAL_CEPH uses (I'm not sure if the knowledge of the client lock owner is part of the solution or if the locks are just known to be held by a particular Ganesha node and trust is placed in the NFSv4 recovery mechanism to reclaim the locks properly.

**Comment by liyinshubyte** [comment]
@ffilz Hi, we want to use some tricky method to recovery lock, such as lock1 is on ganesha1, then ganesha1 crash and restart. In grace period, client1 re-send lock request ganesha1 to recovery lock1. The new lock request should succeed because it's a recovery of lock1,  but we can not identify it's a recovery, so the new lock request failed. So I want to use some identifier not clientid, maybe client hostname + client pid, to identify the lock. Then we find the new lock request has the same client hostname + client pid with existed lock1, so we think it's a recovery of lock1, and respond with success.  But the problem is we can not get client pid which send the lock request, do you know any other way to achieve this ?

**Comment by liyinshubyte** [comment]
> The "theory" is laid out in this slide deck (for the most part):
> 
> 
> [http://nfsv4bat.org/Documents/BakeAThon/2018/Active_Active%20NFS%20Server...](http://nfsv4bat.org/Documents/BakeAThon/2018/Active_Active%20NFS%20Server%20Recovery.pdf)
> 
> Otherwise, when the Ganesha process fails (whether the process itself
> crashes, or it's running on a cluster node that fails), all state held
> by clients is dropped and other processes are free to acquire
> conflicting state without Ganesha necessarily being aware that happened.
> Clustered Ganesha does at least attempt to enter grace period on all
> nodes, however, this is not synchronous with the failure so can leave a
> small window and of course there is no integration with any other
> processes that might be sharing the files (thus some vendors do not
> support multi-protocol, NFS and CIFS for example, sharing of the same
> file sets).

@jtlayton Hi, I learned your shared slides, thanks a lot for the share, but I have a question. For example there are 3 ganesha: ganesha1, ganesha2 and ganesha3. When ganesha1 restart, does all ganesha need to enter grace period,  including ganesha2 and ganesha3 ? But this will cause the cluster unavailable to provide service, and all clients hang, is there other way to improve this ?

@ffilz  Can we change grace period from cluster level to inode level ? For example,  inode1 has lock1 on ganesha1, and inode2 does not has lock. When ganesha1 restart, we respond all request for inode1 with NFS4ERR_GRACE, other than reclaim request for inode1. But the request for inode2 can respond normally. In this way, we avoid make whole cluster enter grace period and unavailable to clients. What do you think ?

**Comment by ffilz** [comment]
Yes, all nodes in a cluster must synchronize grace period.

You can't do grace period per inode because you have no way of knowing for which inodes recovery is complete on.

What you COULD do is persistently track the inodes that have state on them, and allow access to any that didn't have state on them during grace period. The problem is that any inode of interest in NFSv4 has an OPEN on it. You could fine tune the tracking and track what kind of state is on each inode persistently, and track if OPEN state has deny modes or not. Possibly you could even persist the union of the access modes and deny modes to allow non-conflicting opens to occur. But this becomes tricky. In fact even just persisting the existence becomes tricky because if two Ganesha nodes have open state on an inode, how do you know when you can remove the persistent indication? You would end up having a persistent refcount.

**Comment by liyinshubyte** [comment]
@ffilz Thanks the reply, I have another question, sorry for that. 

I use NFSv3 in ganesha, for example ganesha's ip is 1.1.1.1, and in front of ganesha we have a LoadBalance, which has a ip 2.2.2.2, so client use 2.2.2.2 to mount. 

Now client use flock to lock file, then I restart ganesha and execute "sm-notify -f". Then the problem occurs, the notify request's src ip is 1.1.1.1, not 2.2.2.2, so rpc.statd in client will ignore this notify, because it does not match 2.2.2.2, so client will not reclaim lock.

When I did not use LoadBalance, i.e. client use 1.1.1.1 to mount, everything is OK.

Do you know how to make client reclaim lock when use LoadBalance ?

**Comment by ffilz** [comment]
You would need to send sm-notify from the load balancer node. You might be able to send it from the Ganesha node and make it look like it came from the load balancer.

**Comment by liyinshubyte** [comment]
@ffilz Thanks the reply, I have another question, sorry for that. How can I make client reclaim lock when create session, even without ganesha restart ? We need this in the following two cases.

1.  When I use NFSv4 in ganesha, there are two ganesha: ganesha1's ip is ip1, and ganesha2's ip is ip2, and in front of two ganesha we have a LoadBalance, which has a ip, i.e. vip. And client use vip to mount, client's session located on ganesha1. Now client use flock to lock file, then I stop ganesha1, client's request forward to ganesha2 and create new session on ganesha2. But I find client will not reclaim lock on ganesha2.

2.  When there is network problem, for example LoadBalance can not work for a while, client close connection with ganesha1, and re-connect to ganesha2, we also need to recovery lock on ganesha2. In this case, there is no ganesha crash, but we also should make client reclaim lock.

**Comment by liyinshubyte** [comment]
> Possibly you could even persist the union of the access modes and deny modes to allow non-conflicting opens to occur.

@ffilz As the NFSv4 proto describes, only windows have access and deny mode. Does it mean we do not need to persist open state when we only use linux client, and only need to persist lock state ? 

![image](https://user-images.githubusercontent.com/92144382/177720150-20643760-6348-4918-bba7-2f6f62b7f453.png)

**Comment by ffilz** [comment]
> @ffilz Thanks the reply, I have another question, sorry for that. How can I make client reclaim lock when create session, even without ganesha restart ? We need this in the following two cases.
> 
>     1. When I use NFSv4 in ganesha, there are two ganesha: ganesha1's ip is ip1, and ganesha2's ip is ip2, and in front of two
> ganesha we have a LoadBalance, which has a ip, i.e. vip. And client use vip to mount, client's session located on ganesha1. Now
> client use flock to lock file, then I stop ganesha1, client's request forward to ganesha2 and create new session on ganesha2. But I
> find client will not reclaim lock on ganesha2.

The recovery data base needs to be shared among all the Ganesha nodes. See the various recovery data bases in src/SAL/recovery. The recovery data base could use the filesystem version if it is located on the shared filesystem.

>     2. When there is network problem, for example LoadBalance can not work for a while, client close connection with ganesha1,
> and re-connect to ganesha2, we also need to recovery lock on ganesha2. In this case, there is no ganesha crash, but we also
> should make client reclaim lock.

Client reassignment by load balancer with out a node failure requires additional sophistication. The folks at Google are exploring this. One big issue is that ganesha2 has none of the sequence ID, client ID, and state ID information from ganesha1. Additionally "reclaiming" locks without node failure would require some mechanism to drop the locks on ganesha1 atomically with acquiring them on ganesha2. The problems are solvable, but complex. If possible, I suggest you try and join our community calls at 7:00 AM Pacific Time on Tuesdays. I think your needs are in much alignment with Googles.

Ultimately there are a lot of issues resulting from a load balancer hiding the server IP address and connection from the clients. Load balancing using DNS round robin to virtual IP addresses that can be relocated between servers is a much more understood mechanism.

**Comment by ffilz** [comment]
> > Possibly you could even persist the union of the access modes and deny modes to allow non-conflicting opens to occur.
> 
> @ffilz As the NFSv4 proto describes, only windows have access and deny mode. Does it mean we do not need to persist open state when we only use linux client, and only need to persist lock state ?
> 
> ![image](https://user-images.githubusercontent.com/92144382/177720150-20643760-6348-4918-bba7-2f6f62b7f453.png)

Yes, if you don't need to worry about Windows clients you could in theory ignore the implications of deny modes. I'm not sure the complexity this would introduce to the Ganesha code would be worth it though.

**Comment by liyinshubyte** [comment]
>  If possible, I suggest you try and join our community calls at 7:00 AM Pacific Time on Tuesdays. I think your needs are in much alignment with Googles.

@ffilz Thanks, will join it. Another question, is there a way to force client reclaim lock when client create new session ? Maybe we can respond some error to notify client to reclaim lock ?

**Comment by dang** [comment]
That's what Grace is for, to force clients to reclaim locks.

**Comment by ffilz** [comment]
> > If possible, I suggest you try and join our community calls at 7:00 AM Pacific Time on Tuesdays. I think your needs are in much alignment with Googles.
> 
> @ffilz Thanks, will join it. Another question, is there a way to force client reclaim lock when client create new session ? Maybe we can respond some error to notify client to reclaim lock ?

NFS4ERR_STALE_STATEID and NFS4ERR_STALE_CLIENTID indicate a server reboot and need to reclaim state. The problem is the server has to know the presented stateid or clientid belonged to a previous instance as opposed to being completely invalid or expired.

NFS4ERR_GRACE does NOT prompt a client to reclaim. In fact we rely on that when we go into grace on nodes that didn't crash (otherwise we would require ALL state across the cluster be reclaimed).

So possibly with enough information, we could use NFS4ERR_STALE_STATEID or NFS4ERR_STALE_CLIENTID to trigger the client, but we need to persist enough information to validate the stateid or clientid. And then if the situation is because of load balance and not a crash, we need to have a way for the reclaimed state to supersede the state on the Ganesha instance the client was originally using, all without dropping exclusive locks (and deny modes) to avoid exposing the client to effective loss of lock. Delegations and layouts must also be handled if applicable.

**Comment by liyinshubyte** [comment]
> So possibly with enough information, we could use NFS4ERR_STALE_STATEID or NFS4ERR_STALE_CLIENTID to trigger the client, but we need to persist enough information to validate the stateid or clientid. And then if the situation is because of load balance and not a crash, we need to have a way for the reclaimed state to supersede the state on the Ganesha instance the client was originally using, all without dropping exclusive locks (and deny modes) to avoid exposing the client to effective loss of lock. Delegations and layouts must also be handled if applicable.

@ffilz Thanks the suggestion. I find another way to trigger the client to reclaim lock.

When client's connection move from ganesha1 to ganesha2, client will not reclaim lock. The reason is kernel will check server_scope, if server_scope mismatch, client will not reclaim lock, and the server_scope in ganesha is ganesha's hotname, i.e. `char cid_server_owner[MAXNAMLEN+1]`. So if I change `cid_server_owner` from hostname to a common string, such as vip, then client will reclaim lock when moved to ganesha2(ganesha2 need to be in grace period). But I'm not sure whether this will lead to new bug. 

The related code of kernel as follows: kernel will set `NFS4CLNT_SERVER_SCOPE_MISMATCH` when client move from ganesha1 to ganesha2, and will not call `nfs4_state_start_reclaim_nograce` when `NFS4CLNT_SERVER_SCOPE_MISMATCH`
![image](https://user-images.githubusercontent.com/92144382/177897819-89e60f07-b101-41ff-aed6-5c01cde34b49.png)
![image](https://user-images.githubusercontent.com/92144382/177897980-3b0b429a-299d-4b66-8938-71c83a48856c.png)

**Comment by ffilz** [comment]
Sorry I forgot to mention the new NFSV4 { } parameter Server_Scope which is in V4.0 that allows configuring all the nodes with the same server scope. That may be of help.

**Comment by liyinshubyte** [comment]
> Additionally "reclaiming" locks without node failure would require some mechanism to drop the locks on ganesha1 atomically with acquiring them on ganesha2. 

@ffilz Hi, I'm curious about why we need to drop the locks on ganesha1, because no client can use the old lock on ganesha1, maybe there is no problem if we leave it there ?

**Comment by dang** [comment]
ganesha1 needs to drop the lock in the backing store, so that ganesha2 can take it.

**Comment by liyinshubyte** [comment]
@dang Got it, thanks the reply.

**Comment by liyinshubyte** [comment]
@ffilz hi, currently In grace period, when client send setxattr, removexattr, setattr, rename, remove, open and lock request, ganesha will respond with NFS4ERR_GRACE.

Because posix lock is advisory lock. So for linux, we only need to respond lock request with NFS4ERR_GRACE. And for windows, maybe we only need to respond open and lock with NFS4ERR_GRACE.

So why we need to respond setxattr, removexattr, setattr, rename, remove request with NFS4ERR_GRACE ?

**Comment by ffilz** [comment]
From RFC 7530 Section 9.6.2. Server Failure and Recovery:

The period of special handling of locking and READs and WRITEs, equal
in duration to the lease period, is referred to as the "grace
period". During the grace period, clients recover locks and the
associated state by reclaim-type locking requests (i.e., LOCK
requests with reclaim set to TRUE and OPEN operations with a claim
type of either CLAIM_PREVIOUS or CLAIM_DELEGATE_PREV). During the
grace period, the server must reject READ and WRITE operations and
non-reclaim locking requests (i.e., other LOCK and OPEN operations)
with an error of NFS4ERR_GRACE.

If the server can reliably determine that granting a non-reclaim
request will not conflict with reclamation of locks by other clients,
the NFS4ERR_GRACE error does not have to be returned and the
non-reclaim client request can be serviced. For the server to be
able to service READ and WRITE operations during the grace period, it
must again be able to guarantee that no possible conflict could arise
between an impending reclaim locking request and the READ or WRITE
operation. If the server is unable to offer that guarantee, the
NFS4ERR_GRACE error must be returned to the client.

For a server to provide simple, valid handling during the grace
period, the easiest method is to simply reject all non-reclaim
locking requests and READ and WRITE operations by returning the
NFS4ERR_GRACE error. However, a server may keep information about
granted locks in stable storage. With this information, the server
could determine if a regular lock or READ or WRITE operation can be
safely processed.

NFS4ERR_GRACE is also allowed as a return on other operations. For SETATTR which needs to use a stateid4, it definitely makes sense. For other operations, consider if a delegation was outstanding. Without knowing about delegations being reclaimed, the server would not know if it can serve, say a GETATTR, without recalling the delegation.

**Comment by ffilz** [comment]
So certainly we COULD get more smart about when we can reliably avoid returning NFS4ERR_GRACE. More persistent information would help. Knowing we CAN'T hand out layouts and delegations would help.

**Comment by liyinshubyte** [comment]
> If possible, I suggest you try and join our community calls at 7:00 AM Pacific Time on Tuesdays.

@ffilz Hi, what's the link of  community calls ?

**Comment by ffilz** [comment]
> > If possible, I suggest you try and join our community calls at 7:00 AM Pacific Time on Tuesdays.
> 
> @ffilz Hi, what's the link of community calls ?

The information for next week is (the call links should be good for every week):

Ganesha community call
Tuesday, July 19 · 7:00 – 7:30am PDT (UTC - 7)
Google Meet joining info
Video call link: https://meet.google.com/mkh-ctnj-tqz
Or dial: ‪(US) +1 401-702-0462‬ PIN: ‪495 972 631‬#
More phone numbers: https://tel.meet/mkh-ctnj-tqz?pin=2844708468265
Or join via SIP: sip:2844708468265@gmeet.redhat.com

**Comment by ffilz** [comment]
We are considering options for some level of state persistence to allow transparent state migration.
